android/src/main/java/com/jhomlala/better_player/BetterPlayer.java

```diff
@@ -122,6 +125,71 @@
         setupVideoPlayer(eventChannel, textureEntry, result);
     }
 
+    static void preCache(Context context, String key, String dataSource, long preCacheSize, long maxCacheSize, long maxCacheFileSize, Map<String, String> headers, Result result) {
+
+        Uri uri = Uri.parse(dataSource);
+        DataSource.Factory dataSourceFactory;
+
+        String userAgent = System.getProperty(USER_AGENT_PROPERTY);
+        if (headers != null && headers.containsKey(USER_AGENT)) {
+            String userAgentHeader = headers.get(USER_AGENT);
+            if (userAgentHeader != null) {
+                userAgent = userAgentHeader;
+            }
+        }
+
+        if (isHTTP(uri)) {
+            dataSourceFactory = new DefaultHttpDataSource.Factory()
+                    .setUserAgent(userAgent)
+                    .setAllowCrossProtocolRedirects(true)
+                    .setConnectTimeoutMs(DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS)
+                    .setReadTimeoutMs(DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS);
+
+            if (headers != null) {
+                ((DefaultHttpDataSource.Factory) dataSourceFactory).setDefaultRequestProperties(headers);
+            }
+
+
+            DataSpec dataSpec = new DataSpec(uri, 0, preCacheSize);
+
+            class ProgressListener implements CacheWriter.ProgressListener {
+                @Override
+                public void onProgress(long requestLength, long bytesCached, long newBytesCached) {
+                    if (bytesCached >= requestLength) {
+                        //notify the requested size has finished caching
+                        new Handler(Looper.getMainLooper()).post(() -> result.success(null));
```

Comment:
You're creating each time new handler to post event. Why don't you just have one Handler for this?